contrib/mesos/pkg/queue/delay.go

```diff
@@ -353,8 +378,7 @@ func (q *DelayFIFO) pop(cancel <-chan struct{}) interface{} {
 					// we may not have the lock yet, so
 					// broadcast to abort Wait, then
 					// return after lock re-acquisition
-					q.cond().Broadcast()
-					<-signal
+					sitAndSpin(q.cond(), signal)
```

Comment:
it's not safe (for the `cancel` case) to lock before broadcast because if Wait() returns and `cancel` and `signal` are both closed by the time the select executes, `cancel` could win the tie and then deadlock because the lock will already be held.